org/mozilla/jss/ssl/javax/JSSSocketChannel.java

```diff
@@ -0,0 +1,465 @@
+                        // the remote server is really slow.
+                        Thread.sleep(handshakeAttempts * 10);
+                    } catch (Exception e) {}
+                }
+
+                if (handshakeAttempts > maxHandshakeAttempts) {
+                    if (!isBlocking()) {
+                        // In the event we failed to connect under a
+                        // non-blocking socket, return false rather than fail
+                        // here. It could just be that we don't have enough
+                        // data to continue. In that case, doHandshake() in
+                        // JSSSocket will re-try until the connection succeeds.
+                        return false;
+                    }
+
+                    String msg = "Error attempting to handshake: unable to ";
+                    msg += "complete handshake successfully in ";
+                    msg += maxHandshakeAttempts + " calls to wrap or unwrap. ";
+                    msg += "Connection stalled.";
+                    throw new IOException(msg);
+                }
+            } while (state != SSLEngineResult.HandshakeStatus.FINISHED && state != SSLEngineResult.HandshakeStatus.NOT_HANDSHAKING);
+        } catch (SSLException ssle) {
+            String msg = "Error attempting to handshake with remote peer: ";
+            msg += "got unexpected exception: " + ssle.getMessage();
+            throw new IOException(msg, ssle);
+        }
+
+        sslSocket.notifyHandshakeCompletedListeners();
+
+        return true;
+    }
+
+    /**
+     * Compute the total size of a list of buffers from the specified offest
+     * and length.
+     */
+    private static long computeSize(ByteBuffer[] buffers, int offset, int length) throws IOException {
+        long result = 0;
+
+        if (buffers == null || buffers.length == 0) {
+            return result;
+        }
+
+        for (int rel_index = 0; rel_index < length; rel_index++) {
+            int index = offset + rel_index;
+            if (index >= buffers.length) {
+                String msg = "Offset (" + offset + " or length (" + length;
+                msg += ") exceeds contract based on number of buffers ";
+                msg += "given (" + buffers.length + ")";
+                throw new IOException(msg);
+            }
+
+            if (buffers[index] != null) {
+                result += buffers[index].remaining();
+            }
+        }
+
+        return result;
+    }
+
+    public int read(ByteBuffer dst) throws IOException {
+        return (int) read(new ByteBuffer[] { dst });
+    }
+
+    public long read(ByteBuffer[] dsts, int offset, int length) throws IOException {
+        if (inboundClosed) {
+            return -1;
+        }
+
+        int buffer_size = boundRead(JSSEngine.BUFFER_SIZE);
+        ByteBuffer src = ByteBuffer.allocate(buffer_size);
+
+        long remoteRead = readChannel.read(src);
```

Comment:
@edewata Should we move this inside the loop below? In particular, if we read a non-zero amount and then unwrap it all, we could retry the read and potentially get more data. Thoughts? 

Note that we'd have to handle the case where the `consumed` has data left, but we don't currently handle that either way.

It'd make the `read()` code look more like `write()` though.

Thoughts?

Reply:
In particular, to get `consumed` to work, we replace `readChannel` with  `Channels.newChannel(consumed)` and that'll give us something we can `.read(src)` into -- we just need to make sure to check `consumed.available` and limit the size of `buffer_size` above appropriately.

Reply:
I've gotten `consumed` to work but I haven't moved the read inside the thing as it'd involve resizing the `src` buffer afterwards.